File: Physics Simulator/lab/js/sceneGraph/renderVisitor.js
class RenderVisitor {
    constructor(pContext) {
        this.mStack = [];
        this.setContext(pContext);
    }

    getStack() {
        return this.mStack;
    }
    setStack(pStack) {
        this.mStack = pStack;
    }

    getContext() {
        return this.mContext;
    }
    setContext(pContext) {
        this.mContext = pContext;
    }

    visit(pNode) {
        var type = pNode.getType();
        //console.log("type: " + type);
        switch (type) {
            case "group":
                this.visitGroup(pNode);
                break;
            case "polygon":
                this.visitDraw(pNode);
                break;
            case "transform":
                this.visitTransform(pNode);
                 break;
            case "state":
                this.visitState(pNode);
                break;
            case "circle":
                this.visitDraw(pNode);
                break;
        }
    }

    visitGroup(pNode) {
        for (var i = 0; i < pNode.getChildNum(); i += 1) {
            var node = pNode.getSpecificChild(i);
            node.accept(this);
        }
    }

    visitDraw(pNode) {
        //get the top from the stack, set transform with it
        var topStack = this.peekTransform();
        topStack.setTransform(this.getContext());

        //call the draw of the node
        pNode.draw(this.getContext());
    }

    visitTransform(pNode) {
        //add to stack
        this.pushTransform(pNode.getLTM());
        //call children using visitGroup
        this.visitGroup(pNode);
        //remove top of stack
        this.popTransform();
    }

    visitState(pNode) {
        //console.log("state hit");
        var topic = pNode.getTopic();
        var value = pNode.getValue();
        var previousState = this.mContext[topic];
        this.mContext[topic] = value;
        this.visitGroup(pNode);
        this.mContext[topic] = previousState;
    }

    pushTransform(pMatrix) {
        //if there are 0 things in stack// add to stack
        if (this.mStack.length === 0) {
            this.mStack.push(pMatrix);   
        }
        //if not call peektransform(to get top of stack) and multiply old matrix by the new matrix and push it to the stack
        else {
            var topStack = this.peekTransform();
            var newMatrix = topStack.multiply(pMatrix);
            this.mStack.push(newMatrix);
        }
    }

    popTransform() {
        //removes top transform from stack
        this.mStack.pop();
    }

    peekTransform() {
        //get the length of stack minus 1 and return that number element
        var num = this.mStack.length - 1;
        //console.log(num);
        return this.mStack[num];
    }

}

File: 3D Graphics/Labs/ACW/RenderVisitor.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using OpenTK;
using OpenTK.Graphics;
using OpenTK.Graphics.OpenGL;
using Labs.Utility;

namespace Labs.ACW
{
    class RenderVisitor
    {
        Stack<Matrix4> mMatrixStack;
        Stack<int> mShaderStack;
        Stack<int> mTextureStack;

        public RenderVisitor()
        {
            mMatrixStack = new Stack<Matrix4>();
            mShaderStack = new Stack<int>();
            mTextureStack = new Stack<int>();
        }

        public void Visit(SceneGraphNode node)
        {
            Type t = node.GetType();

            if (t.Equals(typeof(SceneGraphNode)))
            {
                VisitNode(node);
            }
            else if (t.Equals(typeof(SceneGraphGroup)))
            {
                VisitGroup((SceneGraphGroup)node);
            }
            else if (t.Equals(typeof(SceneGraphTransformation)))
            {
                VisitTransformation((SceneGraphTransformation)node);
            }
            else if (t.Equals(typeof(SceneGraphShader)))
            {
                VisitShader((SceneGraphShader)node);
            }
            else if (t.Equals(typeof(SceneGraphTexture)))
            {
                VisitTexture((SceneGraphTexture)node);
            }
            else if (t.Equals(typeof(SceneGraphDrawModel)))
            {
                VisitDrawModel((SceneGraphDrawModel)node);
            }
            else if (t.Equals(typeof(SceneGraphDraw)))
            {
                VisitDraw((SceneGraphDraw)node);
            }
            else if (t.Equals(typeof(SceneGraphDrawElements)))
            {
                VisitDrawElements((SceneGraphDrawElements)node);
            }
        }

        private void VisitNode(SceneGraphNode pNode)
        {

        }

        private void VisitGroup(SceneGraphGroup pNode)
        {
            for (int i = 0; i < pNode.GetChildCount(); i += 1)
            {
                pNode.GetChild(i).Accept(this);
            }
        }

        private void VisitTransformation(SceneGraphTransformation pNode)
        {
            //add to stack
            PushTransform(pNode.GetLTM());
            //call children using visitGroup
            VisitGroup(pNode);
            //remove top of stack
            PopTransform();
        }

        private void VisitShader(SceneGraphShader pNode)
        {
            int shaderID = pNode.GetShaderID();
            mShaderStack.Push(shaderID);
            GL.UseProgram(shaderID);
            VisitGroup(pNode);
            mShaderStack.Pop();
            if(mShaderStack.Count > 0)
            {
                GL.UseProgram(mShaderStack.Peek());
            }
        }

        private void VisitTexture(SceneGraphTexture pNode)
        {
            int textureID = pNode.GetTextureID();
            mTextureStack.Push(textureID);
            GL.BindTexture(TextureTarget.Texture2D, textureID);
            VisitGroup(pNode);
            mTextureStack.Pop();
        }

        private void VisitDrawModel(SceneGraphDrawModel pNode)
        {
            int uModel = GL.GetUniformLocation(mShaderStack.Peek(), "uModel");
            Matrix4 matrix = PeekTransform();
            GL.UniformMatrix4(uModel, true, ref matrix);
            pNode.DrawModel();
        }

        private void VisitDraw(SceneGraphDraw pNode)
        {
            int uModel = GL.GetUniformLocation(mShaderStack.Peek(), "uModel");
            Matrix4 matrix = PeekTransform();
            GL.UniformMatrix4(uModel, true, ref matrix);
            pNode.Draw();
        }

        private void VisitDrawElements(SceneGraphDrawElements pNode)
        {
            int uModel = GL.GetUniformLocation(mShaderStack.Peek(), "uModel");
            Matrix4 matrix = PeekTransform();
            GL.UniformMatrix4(uModel, true, ref matrix);
            pNode.DrawElements();
        }

        private void PushTransform(Matrix4 pMatrix)
        {

            if (mMatrixStack.Count == 0)
            {
               mMatrixStack.Push(pMatrix);
            }
            else
            {
                Matrix4 topStack = PeekTransform();
                Matrix4 newMatrix = pMatrix * topStack;
                mMatrixStack.Push(newMatrix);
            }
        }

        private void PopTransform()
        {
            //removes top transform from stack
            mMatrixStack.Pop();
        }

        private Matrix4 PeekTransform()
        {
            return mMatrixStack.Peek();
        }
    }
}